File: xLiAd.DiagnosticLogCenter.Agent/TracingDiagnosticMethod.cs
using AspectCore.Extensions.Reflection;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using xLiAd.DiagnosticLogCenter.Agent.DiagnosticProcessors;

namespace xLiAd.DiagnosticLogCenter.Agent
{
    internal class TracingDiagnosticMethod
    {
        private readonly ITracingDiagnosticProcessor _tracingDiagnosticProcessor;
        private readonly string _diagnosticName;
        private readonly IParameterResolver[] _parameterResolvers;
        private readonly MethodReflector _reflector;

        public TracingDiagnosticMethod(ITracingDiagnosticProcessor tracingDiagnosticProcessor, MethodInfo method,
            string diagnosticName)
        {
            _tracingDiagnosticProcessor = tracingDiagnosticProcessor;
            _reflector = method.GetReflector();
            _diagnosticName = diagnosticName;
            _parameterResolvers = GetParameterResolvers(method).ToArray();
        }

        public void Invoke(string diagnosticName, object value)
        {
            if (_diagnosticName != diagnosticName)
            {
                return;
            }

            var args = new object[_parameterResolvers.Length];
            for (var i = 0; i < _parameterResolvers.Length; i++)
            {
                args[i] = _parameterResolvers[i].Resolve(value);
            }

            _reflector.Invoke(_tracingDiagnosticProcessor, args);
        }

        private static IEnumerable<IParameterResolver> GetParameterResolvers(MethodInfo methodInfo)
        {
            foreach (var parameter in methodInfo.GetParameters())
            {
                var binder = parameter.GetCustomAttribute<ParameterBinder>();
                if (binder != null)
                {
                    if (binder is ObjectAttribute objectBinder)
                    {
                        if (objectBinder.TargetType == null)
                        {
                            objectBinder.TargetType = parameter.ParameterType;
                        }
                    }
                    if (binder is PropertyAttribute propertyBinder)
                    {
                        if (propertyBinder.Name == null)
                        {
                            propertyBinder.Name = parameter.Name;
                        }
                    }
                    yield return binder;
                }
                else
                {
                    yield return new NullParameterResolver();
                }
            }
        }
    }
}

File: DiagnosticLogCenter-Java/src/main/java/com/xliad/diagnosticlogcenter/agent/TracingDiagnosticMethod.java
package com.xliad.diagnosticlogcenter.agent;

import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.IParameterResolver;
import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.ITracingDiagnosticProcessor;
import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.NullParameterResolver;
import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.ParameterBinder;
import com.xliad.diagnosticlogcenter.agent.diagnosticprocessors.PropertyAttribute;
import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.List;

public class TracingDiagnosticMethod {
    private final ITracingDiagnosticProcessor processor;
    private final Method method;
    private final String diagnosticName;
    private final List<IParameterResolver> parameterResolvers = new ArrayList<>();

    public TracingDiagnosticMethod(ITracingDiagnosticProcessor processor, Method method, String diagnosticName) {
        this.processor = processor;
        this.method = method;
        this.diagnosticName = diagnosticName;
        initParameterResolvers();
    }

    private void initParameterResolvers() {
        for (Parameter parameter : method.getParameters()) {
            IParameterResolver resolver = null;

            for (Annotation annotation : parameter.getAnnotations()) {
                if (annotation instanceof ParameterBinder) {
                    resolver = (IParameterResolver) annotation;
                    break;
                }
            }

            if (resolver == null) {
                resolver = new NullParameterResolver();
            }

            parameterResolvers.add(resolver);
        }
    }

    public void invoke(String eventName, Object value) {
        if (!diagnosticName.equals(eventName)) {
            return;
        }

        Object[] args = new Object[parameterResolvers.size()];
        for (int i = 0; i < parameterResolvers.size(); i++) {
            args[i] = parameterResolvers.get(i).resolve(value);
        }

        try {
            method.setAccessible(true);
            method.invoke(processor, args);
        } catch (Exception e) {
            // 忽略异常
        }
    }
}